### DISCUSSION TITLE
Warn user before leaving resource with unsaved changes

### DISCUSSION THREAD

**Discussion by Dontorpedo** [filamentphp/filament#4370] [Ideas]
Functionality like this would be glorious or does filament have a easy way to achieve that?

**Comment by oddvalue** [comment]
+1 on this.
Obviously, it can be hacked in by watching for changes on inputs and triggering a confirm modal using `beforeunload` but it would be nice to have a built-in method.

**Comment by zepfietje** [comment]
I agree this might be a good feature to have built in. Added it to the roadmap as a proposal!

**Reply by intrepidws** [reply]
Did anything like this ever make it in to Filament?

**Reply by danharrin** [reply]
We'd welcome a PR, but we're not sure on the implementation ourselves

**Reply by intrepidws** [reply]
I'm certainly open to submitting a PR. My use case might be slightly different from the OP, but I have a few ideas about how this might be implemented. However, I'm having trouble putting all of the pieces together - perhaps you could help point me in the right direction.

My initial thought it is that I would emit a Livewire event when a form fails validation. I would register an event listener in JavaScript to catch that event and implement some code using the "beforeunload" Javascript event to warn the user when they try to leave the page. When the form eventually passes validation, I would take similar steps but this time remove the "beforeunload" code.

Perhaps Filament already emits a Livewire event when the form fails or passes validation? If not, could you advise on how I might do that when using the code in the [Sending Validation Notifications](https://filamentphp.com/docs/2.x/forms/validation#sending-validation-notifications) section of the documentation?

**Reply by pboivin** [reply]
I'm not sure validation is the right hook for this. I think I would want to show a warning if a user simply types into a field and then tries to close the browser tab.

**Reply by intrepidws** [reply]
Right, as I said my use case might be slightly different. But either way, I think the "afterValidate" hook (which only fires on successful validation) is the correct place to remove the "beforeunload" Javascript event.

I have a proof of concept working on my end that uses both the "beforeValidate" and "afterValidate" hooks, along with a Livewire event listener in JavaScript. So if you wanted to show the warning when any field has changed (regardless of validation) then the question becomes how to replace the "beforeValidation" step for that purpose. Maybe there is some sort of "dirty" state that could be hooked in to?

**Reply by bilogic** [reply]
My form has a dropdown selection and when I change options, a spinner is prefixed to the text.

![image](https://github.com/filamentphp/filament/assets/946010/0b5ac5d8-e26c-4315-bcb8-e8e19fccfa03)

1. When I first saw it, I thought the form was automatically detecting changes and saving itself (And that's really confusing!), but in reality, what it was trying to convey is that, you can't click the `Save` button temporarily (while it's spinning).

2. Whatever that made the button spin could be used to set a `dirty` flag in the `Save` button and only cleared after `Save` is clicked. 

3. When dirty, the `Save` button can have an asterisk or something prepended/appended that makes it look different. The warning can be abstracted away inside the `Save` button's blade.

4. Back to point 1, instead of the spinner, I think it makes more sense to just disable the button such that it cannot be clicked. I contemplated graying it out, but that likely will have the same confusing effect as the spinner. Bottomline, being unable `Save` temporarily is not so critical that it needs to attract the user's attention.

**Comment by danharrin** [comment]
https://filamentphp.com/docs/4.x/panel-configuration#unsaved-changes-alerts
